modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/ContentSecurityPolicyNonceManager.java

```diff
@@ -27,57 +27,64 @@
 @Component(service = ContentSecurityPolicyNonceManager.class)
 public class ContentSecurityPolicyNonceManager {
 
-	public String ensureNonce(HttpServletRequest httpServletRequest) {
+	public void cleanUpNonce(HttpServletRequest httpServletRequest) {
 		httpServletRequest = _portal.getOriginalServletRequest(
 			httpServletRequest);
 
-		HttpSession httpSession = httpServletRequest.getSession();
+		httpServletRequest.removeAttribute(_NONCE);
 
-		String nonce = (String)httpSession.getAttribute(_NONCE);
+		_threadLocal.remove();
+	}
+
+	public String getNonce(HttpServletRequest httpServletRequest) {
+		String nonce = GetterUtil.getString(_threadLocal.get());
 
 		if (nonce != null) {
 			return nonce;
 		}
 
-		synchronized (httpSession) {
-			nonce = (String)httpSession.getAttribute(_NONCE);
-
-			if (nonce == null) {
-				nonce = _generateNonce();
-
-				httpSession.setAttribute(_NONCE, nonce);
-			}
-		}
-
-		return nonce;
+		return GetterUtil.getString(httpServletRequest.getAttribute(_NONCE));
 	}
 
-	public String getNonce(HttpServletRequest httpServletRequest) {
-		if (httpServletRequest == null) {
-			return GetterUtil.getString(_threadLocal.get());
-		}
+	public String setNonce(HttpServletRequest httpServletRequest) {
+		httpServletRequest = _portal.getOriginalServletRequest(
+			httpServletRequest);
+
+		String nonce;
 
 		ContentSecurityPolicyConfiguration contentSecurityPolicyConfiguration =
-			_getContentSecurityPolicyConfiguration(httpServletRequest);
+			ContentSecurityPolicyConfigurationUtil.
+				getContentSecurityPolicyConfiguration(httpServletRequest);
 
 		if (!contentSecurityPolicyConfiguration.enabled()) {
-			return StringPool.BLANK;
+			nonce = StringPool.BLANK;
 		}
+		else if (PropsValues.JAVASCRIPT_SINGLE_PAGE_APPLICATION_ENABLED) {
+			HttpSession httpSession = httpServletRequest.getSession();
 
-		httpServletRequest = _portal.getOriginalServletRequest(
-			httpServletRequest);
+			nonce = (String)httpSession.getAttribute(_NONCE);
```

Comment:
Wouldn't we like to include getting the nonce logic inside the `synchronized` block? This way we won't allow 2 threads to get a null value before trying to set it

Reply:
just realized you did it inside the syncrhonized. So pls disregard my previoius comment

Reply:
:+1: